File: app/src/androidTest/java/com/todoroo/astrid/repeats/NewRepeatTests.java
package com.todoroo.astrid.repeats;

import static com.todoroo.astrid.repeats.RepeatTaskHelper.computeNextDueDate;
import static java.util.Arrays.asList;
import static junit.framework.Assert.assertEquals;

import android.support.test.runner.AndroidJUnit4;
import com.google.ical.values.Frequency;
import com.google.ical.values.RRule;
import com.google.ical.values.Weekday;
import com.google.ical.values.WeekdayNum;
import com.todoroo.astrid.data.Task;
import java.text.ParseException;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.tasks.time.DateTime;

@RunWith(AndroidJUnit4.class)
public class NewRepeatTests {

  @Test
  public void testRepeatMinutelyFromDueDate() throws ParseException {
    DateTime dueDateTime = newDayTime(2016, 8, 26, 12, 30);
    Task task = newFromDue(Frequency.MINUTELY, 1, dueDateTime);

    assertEquals(newDayTime(2016, 8, 26, 12, 31), calculateNextDueDate(task));
  }

  @Test
  public void testRepeatHourlyFromDueDate() throws ParseException {
    DateTime dueDateTime = newDayTime(2016, 8, 26, 12, 30);
    Task task = newFromDue(Frequency.HOURLY, 1, dueDateTime);

    assertEquals(newDayTime(2016, 8, 26, 13, 30), calculateNextDueDate(task));
  }

  @Test
  public void testRepeatDailyFromDueDate() throws ParseException {
    DateTime dueDateTime = newDayTime(2016, 8, 26, 12, 30);
    Task task = newFromDue(Frequency.DAILY, 1, dueDateTime);

    assertEquals(newDayTime(2016, 8, 27, 12, 30), calculateNextDueDate(task));
  }

  @Test
  public void testRepeatWeeklyFromDueDate() throws ParseException {
    DateTime dueDateTime = newDayTime(2016, 8, 28, 1, 34);
    Task task = newFromDue(Frequency.WEEKLY, 1, dueDateTime);

    assertEquals(newDayTime(2016, 9, 4, 1, 34), calculateNextDueDate(task));
  }

  @Test
  public void testRepeatMonthlyFromDueDate() throws ParseException {
    DateTime dueDateTime = newDayTime(2016, 8, 28, 1, 44);
    Task task = newFromDue(Frequency.MONTHLY, 1, dueDateTime);

    assertEquals(newDayTime(2016, 9, 28, 1, 44), calculateNextDueDate(task));
  }

  @Test
  public void testRepeatYearlyFromDueDate() throws ParseException {
    DateTime dueDateTime = newDayTime(2016, 8, 28, 1, 44);
    Task task = newFromDue(Frequency.YEARLY, 1, dueDateTime);

    assertEquals(newDayTime(2017, 8, 28, 1, 44), calculateNextDueDate(task));
  }

  /** Tests for repeating from completionDate */
  @Test
  public void testRepeatMinutelyFromCompleteDateCompleteBefore() throws ParseException {
    DateTime dueDateTime = newDayTime(2016, 8, 30, 0, 25);
    DateTime completionDateTime = newDayTime(2016, 8, 29, 0, 14);
    Task task = newFromCompleted(Frequency.MINUTELY, 1, dueDateTime, completionDateTime);

    assertEquals(newDayTime(2016, 8, 29, 0, 15), calculateNextDueDate(task));
  }

  @Test
  public void testRepeatMinutelyFromCompleteDateCompleteAfter() throws ParseException {
    DateTime dueDateTime = newDayTime(2016, 8, 28, 0, 4);
    DateTime completionDateTime = newDayTime(2016, 8, 29, 0, 14);
    Task task = newFromCompleted(Frequency.MINUTELY, 1, dueDateTime, completionDateTime);

    assertEquals(newDayTime(2016, 8, 29, 0, 15), calculateNextDueDate(task));
  }

  @Test
  public void testRepeatHourlyFromCompleteDateCompleteBefore() throws ParseException {
    DateTime dueDateTime = newDayTime(2016, 8, 30, 0, 25);
    DateTime completionDateTime = newDayTime(2016, 8, 29, 0, 14);
    Task task = newFromCompleted(Frequency.HOURLY, 1, dueDateTime, completionDateTime);

    assertEquals(newDayTime(2016, 8, 29, 1, 14), calculateNextDueDate(task));
  }

  @Test
  public void testRepeatHourlyFromCompleteDateCompleteAfter() throws ParseException {
    DateTime dueDateTime = newDayTime(2016, 8, 28, 0, 4);
    DateTime completionDateTime = newDayTime(2016, 8, 29, 0, 14);
    Task task = newFromCompleted(Frequency.HOURLY, 1, dueDateTime, completionDateTime);

    assertEquals(newDayTime(2016, 8, 29, 1, 14), calculateNextDueDate(task));
  }

  @Test
  public void testRepeatDailyFromCompleteDateCompleteBefore() throws ParseException {
    DateTime dueDateTime = newDayTime(2016, 8, 30, 0, 25);
    DateTime completionDateTime = newDayTime(2016, 8, 29, 0, 14);
    Task task = newFromCompleted(Frequency.DAILY, 1, dueDateTime, completionDateTime);

    assertEquals(newDayTime(2016, 8, 30, 0, 25), calculateNextDueDate(task));
  }

  @Test
  public void testRepeatDailyFromCompleteDateCompleteAfter() throws ParseException {
    DateTime dueDateTime = newDayTime(2016, 8, 28, 0, 4);
    DateTime completionDateTime = newDayTime(2016, 8, 29, 0, 14);
    Task task = newFromCompleted(Frequency.DAILY, 1, dueDateTime, completionDateTime);

    assertEquals(newDayTime(2016, 8, 30, 0, 4), calculateNextDueDate(task));
  }

  @Test
  public void testRepeatWeeklyFromCompleteDateCompleteBefore() throws ParseException {
    DateTime dueDateTime = newDayTime(2016, 8, 30, 0, 25);
    DateTime completionDateTime = newDayTime(2016, 8, 29, 0, 14);
    Task task = newFromCompleted(Frequency.WEEKLY, 1, dueDateTime, completionDateTime);

    assertEquals(newDayTime(2016, 9, 5, 0, 25), calculateNextDueDate(task));
  }

  @Test
  public void testRepeatWeeklyFromCompleteDateCompleteAfter() throws ParseException {
    DateTime dueDateTime = newDayTime(2016, 8, 28, 0, 4);
    DateTime completionDateTime = newDayTime(2016, 8, 29, 0, 14);
    Task task = newFromCompleted(Frequency.WEEKLY, 1, dueDateTime, completionDateTime);

    assertEquals(newDayTime(2016, 9, 5, 0, 4), calculateNextDueDate(task));
  }

  @Test
  public void testRepeatMonthlyFromCompleteDateCompleteBefore() throws ParseException {
    DateTime dueDateTime = newDayTime(2016, 8, 30, 0, 25);
    DateTime completionDateTime = newDayTime(2016, 8, 29, 0, 14);
    Task task = newFromCompleted(Frequency.MONTHLY, 1, dueDateTime, completionDateTime);

    assertEquals(newDayTime(2016, 9, 29, 0, 25), calculateNextDueDate(task));
  }

  @Test
  public void testRepeatMonthlyFromCompleteDateCompleteAfter() throws ParseException {
    DateTime dueDateTime = newDayTime(2016, 8, 28, 0, 4);
    DateTime completionDateTime = newDayTime(2016, 8, 29, 0, 14);
    Task task = newFromCompleted(Frequency.MONTHLY, 1, dueDateTime, completionDateTime);

    assertEquals(newDayTime(2016, 9, 29, 0, 4), calculateNextDueDate(task));
  }

  @Test
  public void testRepeatYearlyFromCompleteDateCompleteBefore() throws ParseException {
    DateTime dueDateTime = newDayTime(2016, 8, 30, 0, 25);
    DateTime completionDateTime = newDayTime(2016, 8, 29, 0, 14);
    Task task = newFromCompleted(Frequency.YEARLY, 1, dueDateTime, completionDateTime);

    assertEquals(newDayTime(2017, 8, 29, 0, 25), calculateNextDueDate(task));
  }

  @Test
  public void testRepeatYearlyFromCompleteDateCompleteAfter() throws ParseException {
    DateTime dueDateTime = newDayTime(2016, 8, 28, 0, 4);
    DateTime completionDateTime = newDayTime(2016, 8, 29, 0, 14);
    Task task = newFromCompleted(Frequency.YEARLY, 1, dueDateTime, completionDateTime);

    assertEquals(newDayTime(2017, 8, 29, 0, 4), calculateNextDueDate(task));
  }

  @Test
  public void testAdvancedRepeatWeeklyFromDueDate() throws ParseException {
    DateTime dueDateTime = newDayTime(2016, 8, 29, 0, 25);
    Task task =
        newWeeklyFromDue(
            1, dueDateTime, new WeekdayNum(0, Weekday.MO), new WeekdayNum(0, Weekday.WE));

    assertEquals(newDayTime(2016, 8, 31, 0, 25), calculateNextDueDate(task));
  }

  @Test
  public void testAdvancedRepeatWeeklyFromCompleteDateCompleteBefore() throws ParseException {
    DateTime dueDateTime = newDayTime(2016, 8, 29, 0, 25);
    DateTime completionDateTime = newDayTime(2016, 8, 28, 1, 9);
    Task task =
        newWeeklyFromCompleted(
            1,
            dueDateTime,
            completionDateTime,
            new WeekdayNum(0, Weekday.MO),
            new WeekdayNum(0, Weekday.WE));

    assertEquals(newDayTime(2016, 8, 29, 0, 25), calculateNextDueDate(task));
  }

  @Test
  public void testAdvancedRepeatWeeklyFromCompleteDateCompleteAfter() throws ParseException {
    DateTime dueDateTime = newDayTime(2016, 8, 29, 0, 25);
    DateTime completionDateTime = newDayTime(2016, 9, 1, 1, 9);
    Task task =
        newWeeklyFromCompleted(
            1,
            dueDateTime,
            completionDateTime,
            new WeekdayNum(0, Weekday.MO),
            new WeekdayNum(0, Weekday.WE));

    assertEquals(newDayTime(2016, 9, 5, 0, 25), calculateNextDueDate(task));
  }

  private DateTime newDayTime(int year, int month, int day, int hour, int minute) {
    return new DateTime(
        Task.createDueDate(
            Task.URGENCY_SPECIFIC_DAY_TIME,
            new DateTime(year, month, day, hour, minute).getMillis()));
  }

  private DateTime calculateNextDueDate(Task task) throws ParseException {
    return new DateTime(
        computeNextDueDate(task, task.sanitizedRecurrence(), task.repeatAfterCompletion()));
  }

  private Task newFromDue(Frequency frequency, int interval, DateTime dueDateTime) {
    return new Task() {
      {
        setRecurrence(getRecurrenceRule(frequency, interval, false));
        setDueDate(dueDateTime.getMillis());
      }
    };
  }

  private Task newWeeklyFromDue(int interval, DateTime dueDateTime, WeekdayNum... weekdays) {
    return new Task() {
      {
        setRecurrence(getRecurrenceRule(Frequency.WEEKLY, interval, false, weekdays));
        setDueDate(dueDateTime.getMillis());
      }
    };
  }

  private Task newFromCompleted(
      Frequency frequency, int interval, DateTime dueDateTime, DateTime completionDate) {
    return new Task() {
      {
        setRecurrence(getRecurrenceRule(frequency, interval, true));
        setDueDate(dueDateTime.getMillis());
        setCompletionDate(completionDate.getMillis());
      }
    };
  }

  private Task newWeeklyFromCompleted(
      int interval, DateTime dueDateTime, DateTime completionDate, WeekdayNum... weekdays) {
    return new Task() {
      {
        setRecurrence(getRecurrenceRule(Frequency.WEEKLY, interval, true, weekdays));
        setDueDate(dueDateTime.getMillis());
        setCompletionDate(completionDate.getMillis());
      }
    };
  }

  private String getRecurrenceRule(
      Frequency frequency, int interval, boolean fromCompletion, WeekdayNum... weekdays) {
    RRule rrule = new RRule();
    rrule.setFreq(frequency);
    rrule.setInterval(interval);
    if (weekdays != null) {
      rrule.setByDay(asList(weekdays));
    }
    String result = rrule.toIcal();
    if (fromCompletion) {
      result += ";FROM=COMPLETION";
    }
    return result;
  }
}

File: app/src/test/java/com/todoroo/astrid/repeats/NewRepeatTests.kt
package com.todoroo.astrid.repeats

import com.natpryce.makeiteasy.MakeItEasy.with
import com.todoroo.astrid.data.Task
import net.fortuna.ical4j.model.Recur.Frequency
import net.fortuna.ical4j.model.WeekDay
import org.junit.Assert.assertEquals
import org.junit.Test
import org.tasks.makers.TaskMaker.AFTER_COMPLETE
import org.tasks.makers.TaskMaker.COMPLETION_TIME
import org.tasks.makers.TaskMaker.DUE_TIME
import org.tasks.makers.TaskMaker.RECUR
import org.tasks.makers.TaskMaker.newTask
import org.tasks.repeats.RecurrenceUtils.newRecur
import org.tasks.time.DateTime
import java.text.ParseException

class NewRepeatTests {
    @Test
    @Throws(ParseException::class)
    fun testRepeatMinutelyFromDueDate() {
        val dueDateTime = newDayTime(2016, 8, 26, 12, 30)
        val task = newFromDue(Frequency.MINUTELY, 1, dueDateTime)
        assertEquals(newDayTime(2016, 8, 26, 12, 31), calculateNextDueDate(task))
    }

    @Test
    @Throws(ParseException::class)
    fun testRepeatHourlyFromDueDate() {
        val dueDateTime = newDayTime(2016, 8, 26, 12, 30)
        val task = newFromDue(Frequency.HOURLY, 1, dueDateTime)
        assertEquals(newDayTime(2016, 8, 26, 13, 30), calculateNextDueDate(task))
    }

    @Test
    @Throws(ParseException::class)
    fun testRepeatDailyFromDueDate() {
        val dueDateTime = newDayTime(2016, 8, 26, 12, 30)
        val task = newFromDue(Frequency.DAILY, 1, dueDateTime)
        assertEquals(newDayTime(2016, 8, 27, 12, 30), calculateNextDueDate(task))
    }

    @Test
    @Throws(ParseException::class)
    fun testRepeatWeeklyFromDueDate() {
        val dueDateTime = newDayTime(2016, 8, 28, 1, 34)
        val task = newFromDue(Frequency.WEEKLY, 1, dueDateTime)
        assertEquals(newDayTime(2016, 9, 4, 1, 34), calculateNextDueDate(task))
    }

    @Test
    @Throws(ParseException::class)
    fun testRepeatMonthlyFromDueDate() {
        val dueDateTime = newDayTime(2016, 8, 28, 1, 44)
        val task = newFromDue(Frequency.MONTHLY, 1, dueDateTime)
        assertEquals(newDayTime(2016, 9, 28, 1, 44), calculateNextDueDate(task))
    }

    @Test
    @Throws(ParseException::class)
    fun testRepeatYearlyFromDueDate() {
        val dueDateTime = newDayTime(2016, 8, 28, 1, 44)
        val task = newFromDue(Frequency.YEARLY, 1, dueDateTime)
        assertEquals(newDayTime(2017, 8, 28, 1, 44), calculateNextDueDate(task))
    }

    /** Tests for repeating from completionDate  */
    @Test
    @Throws(ParseException::class)
    fun testRepeatMinutelyFromCompleteDateCompleteBefore() {
        val dueDateTime = newDayTime(2016, 8, 30, 0, 25)
        val completionDateTime = newDayTime(2016, 8, 29, 0, 14)
        val task = newFromCompleted(Frequency.MINUTELY, 1, dueDateTime, completionDateTime)
        assertEquals(newDayTime(2016, 8, 29, 0, 15), calculateNextDueDate(task))
    }

    @Test
    @Throws(ParseException::class)
    fun testRepeatMinutelyFromCompleteDateCompleteAfter() {
        val dueDateTime = newDayTime(2016, 8, 28, 0, 4)
        val completionDateTime = newDayTime(2016, 8, 29, 0, 14)
        val task = newFromCompleted(Frequency.MINUTELY, 1, dueDateTime, completionDateTime)
        assertEquals(newDayTime(2016, 8, 29, 0, 15), calculateNextDueDate(task))
    }

    @Test
    @Throws(ParseException::class)
    fun testRepeatHourlyFromCompleteDateCompleteBefore() {
        val dueDateTime = newDayTime(2016, 8, 30, 0, 25)
        val completionDateTime = newDayTime(2016, 8, 29, 0, 14)
        val task = newFromCompleted(Frequency.HOURLY, 1, dueDateTime, completionDateTime)
        assertEquals(newDayTime(2016, 8, 29, 1, 14), calculateNextDueDate(task))
    }

    @Test
    @Throws(ParseException::class)
    fun testRepeatHourlyFromCompleteDateCompleteAfter() {
        val dueDateTime = newDayTime(2016, 8, 28, 0, 4)
        val completionDateTime = newDayTime(2016, 8, 29, 0, 14)
        val task = newFromCompleted(Frequency.HOURLY, 1, dueDateTime, completionDateTime)
        assertEquals(newDayTime(2016, 8, 29, 1, 14), calculateNextDueDate(task))
    }

    @Test
    @Throws(ParseException::class)
    fun testRepeatDailyFromCompleteDateCompleteBefore() {
        val dueDateTime = newDayTime(2016, 8, 30, 0, 25)
        val completionDateTime = newDayTime(2016, 8, 29, 0, 14)
        val task = newFromCompleted(Frequency.DAILY, 1, dueDateTime, completionDateTime)
        assertEquals(newDayTime(2016, 8, 30, 0, 25), calculateNextDueDate(task))
    }

    @Test
    @Throws(ParseException::class)
    fun testRepeatDailyFromCompleteDateCompleteAfter() {
        val dueDateTime = newDayTime(2016, 8, 28, 0, 4)
        val completionDateTime = newDayTime(2016, 8, 29, 0, 14)
        val task = newFromCompleted(Frequency.DAILY, 1, dueDateTime, completionDateTime)
        assertEquals(newDayTime(2016, 8, 30, 0, 4), calculateNextDueDate(task))
    }

    @Test
    @Throws(ParseException::class)
    fun testRepeatWeeklyFromCompleteDateCompleteBefore() {
        val dueDateTime = newDayTime(2016, 8, 30, 0, 25)
        val completionDateTime = newDayTime(2016, 8, 29, 0, 14)
        val task = newFromCompleted(Frequency.WEEKLY, 1, dueDateTime, completionDateTime)
        assertEquals(newDayTime(2016, 9, 5, 0, 25), calculateNextDueDate(task))
    }

    @Test
    @Throws(ParseException::class)
    fun testRepeatWeeklyFromCompleteDateCompleteAfter() {
        val dueDateTime = newDayTime(2016, 8, 28, 0, 4)
        val completionDateTime = newDayTime(2016, 8, 29, 0, 14)
        val task = newFromCompleted(Frequency.WEEKLY, 1, dueDateTime, completionDateTime)
        assertEquals(newDayTime(2016, 9, 5, 0, 4), calculateNextDueDate(task))
    }

    @Test
    @Throws(ParseException::class)
    fun testRepeatMonthlyFromCompleteDateCompleteBefore() {
        val dueDateTime = newDayTime(2016, 8, 30, 0, 25)
        val completionDateTime = newDayTime(2016, 8, 29, 0, 14)
        val task = newFromCompleted(Frequency.MONTHLY, 1, dueDateTime, completionDateTime)
        assertEquals(newDayTime(2016, 9, 29, 0, 25), calculateNextDueDate(task))
    }

    @Test
    @Throws(ParseException::class)
    fun testRepeatMonthlyFromCompleteDateCompleteAfter() {
        val dueDateTime = newDayTime(2016, 8, 28, 0, 4)
        val completionDateTime = newDayTime(2016, 8, 29, 0, 14)
        val task = newFromCompleted(Frequency.MONTHLY, 1, dueDateTime, completionDateTime)
        assertEquals(newDayTime(2016, 9, 29, 0, 4), calculateNextDueDate(task))
    }

    @Test
    @Throws(ParseException::class)
    fun testRepeatYearlyFromCompleteDateCompleteBefore() {
        val dueDateTime = newDayTime(2016, 8, 30, 0, 25)
        val completionDateTime = newDayTime(2016, 8, 29, 0, 14)
        val task = newFromCompleted(Frequency.YEARLY, 1, dueDateTime, completionDateTime)
        assertEquals(newDayTime(2017, 8, 29, 0, 25), calculateNextDueDate(task))
    }

    @Test
    @Throws(ParseException::class)
    fun testRepeatYearlyFromCompleteDateCompleteAfter() {
        val dueDateTime = newDayTime(2016, 8, 28, 0, 4)
        val completionDateTime = newDayTime(2016, 8, 29, 0, 14)
        val task = newFromCompleted(Frequency.YEARLY, 1, dueDateTime, completionDateTime)
        assertEquals(newDayTime(2017, 8, 29, 0, 4), calculateNextDueDate(task))
    }

    @Test
    @Throws(ParseException::class)
    fun testAdvancedRepeatWeeklyFromDueDate() {
        val dueDateTime = newDayTime(2016, 8, 29, 0, 25)
        val task = newWeeklyFromDue(
                1, dueDateTime, WeekDay(WeekDay.MO, 0), WeekDay(WeekDay.WE, 0))
        assertEquals(newDayTime(2016, 8, 31, 0, 25), calculateNextDueDate(task))
    }

    @Test
    @Throws(ParseException::class)
    fun testAdvancedRepeatWeeklyFromCompleteDateCompleteBefore() {
        val dueDateTime = newDayTime(2016, 8, 29, 0, 25)
        val completionDateTime = newDayTime(2016, 8, 28, 1, 9)
        val task = newWeeklyFromCompleted(
                1,
                dueDateTime,
                completionDateTime,
                WeekDay(WeekDay.MO, 0),
                WeekDay(WeekDay.WE, 0))
        assertEquals(newDayTime(2016, 8, 29, 0, 25), calculateNextDueDate(task))
    }

    @Test
    @Throws(ParseException::class)
    fun testAdvancedRepeatWeeklyFromCompleteDateCompleteAfter() {
        val dueDateTime = newDayTime(2016, 8, 29, 0, 25)
        val completionDateTime = newDayTime(2016, 9, 1, 1, 9)
        val task = newWeeklyFromCompleted(
                1,
                dueDateTime,
                completionDateTime,
                WeekDay(WeekDay.MO, 0),
                WeekDay(WeekDay.WE, 0))
        assertEquals(newDayTime(2016, 9, 5, 0, 25), calculateNextDueDate(task))
    }

    private fun newDayTime(year: Int, month: Int, day: Int, hour: Int, minute: Int): DateTime {
        return DateTime(
                Task.createDueDate(
                        Task.URGENCY_SPECIFIC_DAY_TIME,
                        DateTime(year, month, day, hour, minute).millis))
    }

    @Throws(ParseException::class)
    private fun calculateNextDueDate(task: Task): DateTime {
        return DateTime(
                RepeatTaskHelper.computeNextDueDate(task, task.recurrence!!, task.repeatAfterCompletion()))
    }

    private fun newFromDue(frequency: Frequency, interval: Int, dueDateTime: DateTime): Task {
        return newTask(
                with(RECUR, getRecurrenceRule(frequency, interval)),
                with(AFTER_COMPLETE, false),
                with(DUE_TIME, dueDateTime))
    }

    private fun newWeeklyFromDue(interval: Int, dueDateTime: DateTime, vararg weekdays: WeekDay): Task {
        return newTask(
                with(RECUR, getRecurrenceRule(Frequency.WEEKLY, interval, *weekdays)),
                with(AFTER_COMPLETE, false),
                with(DUE_TIME, dueDateTime))
    }

    private fun newFromCompleted(
            frequency: Frequency, interval: Int, dueDateTime: DateTime, completionDate: DateTime): Task {
        return newTask(
                with(RECUR, getRecurrenceRule(frequency, interval)),
                with(AFTER_COMPLETE, true),
                with(DUE_TIME, dueDateTime),
                with(COMPLETION_TIME, completionDate))
    }

    private fun newWeeklyFromCompleted(
            interval: Int, dueDateTime: DateTime, completionDate: DateTime, vararg weekdays: WeekDay): Task {
        return newTask(
                with(RECUR, getRecurrenceRule(Frequency.WEEKLY, interval, *weekdays)),
                with(AFTER_COMPLETE, true),
                with(DUE_TIME, dueDateTime),
                with(COMPLETION_TIME, completionDate))
    }

    private fun getRecurrenceRule(
            frequency: Frequency, interval: Int, vararg weekdays: WeekDay): String {
        val rrule = newRecur()
        rrule.setFrequency(frequency.name)
        rrule.interval = interval
        if (weekdays.isNotEmpty()) {
            rrule.dayList.addAll(weekdays)
        }
        return rrule.toString()
    }
}